### PR TITLE
quiche: handle stream blockage during decodeHeaders and decodeTrailers

### PR DESCRIPTION
**Pull request by danzh2010**
Commit Message: use posted callback to block/unblock quic stream in EnvoyQuicStream::readDisable() with weak_ptr to stream to handle stream life time issue.
Additional Description: currently if readDisabled() is called in decodeHeaders|Trailers(), the stream will be blocked right away which breaks the assumption QUICHE has that a stream shouldn't be blocked during OnInitialHeadersComplete() and OnTrailingHeadersComplete(). 
This change makes the stream state change completely outside of QUICHE call stack. (The unblocking is already outside of QUICHE call stack in existing implementation.) Also simplify the blockage state change logic.

Risk Level: low
Testing: added more unit tests and enabled Http2UpstreamIntegrationTest::ManyLargeSimultaneousRequestWithBufferLimits which was flaky.

Part of #2557 #14829

### CODE REVIEW

File: source/common/quic/envoy_quic_client_stream.cc
Comment (alyssawilk) [inline source/common/quic/envoy_quic_client_stream.cc:120]: looking at SetBlockedUntilFlush it's pretty clear calling it extra times won't waste CPU or be unsafe.  If we read disable then read enable, we may end up calling some spurious SetUnblocked.  That _looks_ ok but can you confirm, and comment in code if it's fine?

**Review by alyssawilk — COMMENTED** [review]
Thanks for tracking this one down!

File: source/common/quic/envoy_quic_stream.h
Comment (antoniovicente) [reply source/common/quic/envoy_quic_stream.h:81]: +1

Please consider use of a SchedulableCallback owned by the stream.  Call scheduleCallbackNextIteration() to schedule the callback; call cancel() or delete the SchedulableCallback to avoid getting called once you have entered the stream termination sequence.

File: source/common/quic/envoy_quic_stream.h
Comment (antoniovicente) [inline source/common/quic/envoy_quic_stream.h:84]: Please use scheduleCallbackNextIteration

Use of scheduleCallbackCurrentIteration can result in an infinite chain of callbacks which starve every other connection in the proxy.

File: source/common/quic/envoy_quic_stream.h
Comment (danzh2010) [reply source/common/quic/envoy_quic_stream.h]: Switched to use SchedulableCallback. Instead of scheduleCallbackNextIteration(), I used scheduleCallbackInCurrentIteration() because all we need is to not call stream SetUnblocked() in place.

File: source/common/quic/envoy_quic_client_stream.cc
Comment (danzh2010) [reply source/common/quic/envoy_quic_client_stream.cc:120]: Yes, an extra OnDataAvailable() won't cause another push as data should have already been pushed in QUICHE stack or remain blocked by other condition.

File: test/common/quic/envoy_quic_client_stream_test.cc
Comment (DavidSchinazi) [inline test/common/quic/envoy_quic_client_stream_test.cc:591]: nit: `if (quic_version_.UsesHttp3()) {`

**Review by DavidSchinazi — DISMISSED** [review]
Nice! Feel free to ignore the nit if that's the only change before submission to avoid having to rerun CI just for that.

File: source/common/quic/envoy_quic_client_stream.cc
Comment (alyssawilk) [reply source/common/quic/envoy_quic_client_stream.cc:120]: SG, can you please comment that in code then?  The other Envoy read enable/disable need parity (one unblock per block) so I think it's worth calling out that this is different.

**Review by alyssawilk — COMMENTED** [review]
LGTM modulo one comment nit

File: source/common/quic/envoy_quic_stream.h
Comment (danzh2010) [reply source/common/quic/envoy_quic_stream.h]: done

File: source/common/quic/envoy_quic_client_stream.cc
Comment (danzh2010) [reply source/common/quic/envoy_quic_client_stream.cc:120]: Commented at scheduling async_stream_blockage_change_ in readDisable() about the transient block and unblock.

File: test/common/quic/envoy_quic_client_stream_test.cc
Comment (danzh2010) [reply test/common/quic/envoy_quic_client_stream_test.cc:591]: done

File: source/common/quic/envoy_quic_client_stream.cc
Comment (alyssawilk) [reply source/common/quic/envoy_quic_client_stream.cc:120]: ah, that's not quite what I was looking for.
Can we do something like comment in switchStreamBlockState that unlike the envoy readDisable calls, that if the underlying QUIC code has received N block calls, it will still unblock on one block call (so parity between block and unblock calls is not necessary)?

Comment (danzh2010) [reply source/common/quic/envoy_quic_client_stream.cc:126]: done

**Review by alyssawilk — APPROVED** [review]
Awesome, thanks for your patience here :-)

**Pull request merged**
